src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,381 @@
+/*
+ * This file is part of AtomVM.
+ *
+ * Copyright 2024 Davide Bettio <davide@uninstall.it>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+ */
+
+#include <sdkconfig.h>
+#ifdef CONFIG_AVM_ENABLE_STORAGE_NIFS
+
+#include <atom.h>
+#include <defaultatoms.h>
+#include <erl_nif.h>
+#include <erl_nif_priv.h>
+#include <interop.h>
+#include <memory.h>
+#include <nifs.h>
+#include <smp.h>
+#include <term.h>
+
+#include "esp32_sys.h"
+
+#include <stdlib.h>
+
+#include <driver/sdmmc_host.h>
+#include <driver/sdspi_host.h>
+#include <esp_vfs_fat.h>
+
+#include <trace.h>
+
+#include "spi_driver.h"
+
+#define TAG "storage_nif"
+
+#ifndef AVM_NO_SMP
+#define SMP_LOCK_INIT(mounted_fs) smp_spinlock_init(&mounted_fs->lock)
+#define SMP_LOCK(mounted_fs) smp_spinlock_lock(&mounted_fs->lock)
+#define SMP_UNLOCK(mounted_fs) smp_spinlock_unlock(&mounted_fs->lock)
+#else
+#define SMP_LOCK_INIT(mounted_fs)
+#define SMP_LOCK(mounted_fs)
+#define SMP_UNLOCK(mounted_fs)
+#endif
+
+// TODO: allow ro option
+enum mount_type
+{
+    Unmounted,
+    FATSPIFlash,
+    FATSDSPI,
+    FATSDMMC
+};
+
+struct MountedFS
+{
+#ifndef AVM_NO_SMP
+    SpinLock lock;
+#endif
+    char *base_path;
+    enum mount_type mount_type;
+    union
+    {
+        sdmmc_card_t *card;
+        wl_handle_t wl;
```

Comment:
I'm curious why not use wear leveling for all mounts? It is supported on SD and flash, for both SPIFFS and FatFS.

It's not in the scope of this PR obviously, but I think it would be very beneficial to explore using a filesystem and wear leveling for the application partition, so we do not prematurely wear out the first few sectors of the app partition from heavy rewrites of the same few blocks at the beginning of the partition.

Reply:
it depends on the kind of usage you are planning.
However I would rather iterate.